modules/frontend/querysharding.go

```diff
@@ -167,3 +155,49 @@ func DoRequests(ctx context.Context, downstream Handler, reqs []*http.Request) (
 
 	return resps, firstErr
 }
+
+func mergeResponses(rrs []RequestResponse) (*http.Response, error) {
+	var errCode = http.StatusOK
+	var errBody io.ReadCloser
+	var combinedTrace []byte
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			body, err := ioutil.ReadAll(rr.Response.Body)
+			rr.Response.Body.Close()
+			if err != nil {
+				return nil, errors.Wrap(err, "error reading response body at query frontend")
+			}
+
+			if len(combinedTrace) == 0 {
+				combinedTrace = body
+			} else {
+				combinedTrace = util.CombineTraces(combinedTrace, body)
+			}
+		} else if rr.Response.StatusCode != http.StatusNotFound {
+			errCode = rr.Response.StatusCode
+			errBody = rr.Response.Body
+		}
+	}
+
+	if errCode == http.StatusOK {
+		if len(combinedTrace) > 0 {
```

Comment:
It's a little weird to conflate len == 0 with 404.  I don't think it's technically possible to push a 0 byte object into tempo but something to keep in mind.

Please add a short comment above this line that acknowledges this and says it's impossible to push a 0 byte trace so this logic is valid.

Reply:
Agree with this. I've added a counter for the 404s. Removing any ambiguity.